prov/verbs/src/fi_verbs.h

```diff
@@ -394,33 +443,181 @@ void fi_ibv_monitor_unsubscribe(struct ofi_mem_monitor *notifier,
 				struct ofi_subscription *subscription);
 struct ofi_subscription *fi_ibv_monitor_get_event(struct ofi_mem_monitor *notifier);
 
+/*
+ * An XRC SRQ cannot be created until the associated RX CQ is known,
+ * maintain a list of validated pre-posted receives to post once
+ * the SRQ is created.
+ */
+struct fi_ibv_xrc_srx_prepost {
+	struct slist_entry	prepost_entry;
+	void			*buf;
+	void			*desc;
+	void			*context;
+	size_t			len;
+	fi_addr_t		src_addr;
+};
+
 struct fi_ibv_srq_ep {
 	struct fid_ep		ep_fid;
 	struct ibv_srq		*srq;
+	struct fi_ibv_domain	*domain;
+
+	/* Due to XRC SRQ semantics, an XRC SRX context may only be bound
+	 * to multiple endpoints when they share the same RX CQ. */
+	struct fi_ibv_cq	*srq_cq;
+
+	/* XRC SRQ is not created until endpoint enable */
+	fastlock_t		prepost_lock;
+	ofi_fastlock_acquire_t	pp_fastlock_acquire;
+	ofi_fastlock_release_t	pp_fastlock_release;
+
+	struct slist		prepost_list;
+	uint32_t		max_recv_wr;
+	uint32_t		max_sge;
+	uint32_t		prepost_count;
 };
 
 int fi_ibv_srq_context(struct fid_domain *domain, struct fi_rx_attr *attr,
 		       struct fid_ep **rx_ep, void *context);
 
+/* Used to determine if XRC has been enabled and requested */
+#ifdef INCLUDE_VERBS_XRC
+static inline int fi_ibv_using_xrc(void)
+{
+	return fi_ibv_gl_data.msg.use_xrc;
+}
+
+#else /* INCLUDE_VERBS_XRC */
+static inline int fi_ibv_using_xrc(void)
+{
+	return 0;
+}
+#endif /* INCLUDE_VERBS_XRC */
+
+int fi_ibv_domain_xrc_init(struct fi_ibv_domain *domain);
+int fi_ibv_domain_xrc_cleanup(struct fi_ibv_domain *domain);
+
+/* Verbs-DGRAM Pool functionality */
```

Comment:
Looks like the DGRAM buffer pool has been removed from master. I'll need to delete the following references to dgram_buf_pool.